### ISSUE TITLE
After using peft_model.merge_and_unload(), the merged model is completely different from the inference result of peft_model.

### ISSUE THREAD

**Issue by weiminw** [huggingface/peft#2502]
### System Info

Package                   Version
------------------------- --------------
absl-py                   2.2.1
accelerate                1.5.2
aiohappyeyeballs          2.6.1
aiohttp                   3.11.15
aiosignal                 1.3.2
anyio                     4.9.0
argon2-cffi               23.1.0
argon2-cffi-bindings      21.2.0
arrow                     1.3.0
asttokens                 3.0.0
async-lru                 2.0.5
async-timeout             5.0.1
attrs                     25.3.0
babel                     2.17.0
beautifulsoup4            4.13.3
bitsandbytes              0.45.4
bleach                    6.2.0
certifi                   2025.1.31
cffi                      1.17.1
charset-normalizer        3.4.1
comm                      0.2.2
contourpy                 1.3.1
cut-cross-entropy         25.1.1
cycler                    0.12.1
datasets                  3.5.0
debugpy                   1.8.13
decorator                 5.2.1
defusedxml                0.7.1
diffusers                 0.32.2
dill                      0.3.8
docstring_parser          0.16
einops                    0.8.1
exceptiongroup            1.2.2
executing                 2.2.0
fastjsonschema            2.21.1
filelock                  3.18.0
flash_attn                2.7.4.post1
fonttools                 4.56.0
fqdn                      1.5.1
frozenlist                1.5.0
fsspec                    2024.12.0
grpcio                    1.71.0
h11                       0.14.0
hf_transfer               0.1.9
httpcore                  1.0.7
httpx                     0.28.1
huggingface-hub           0.30.1
idna                      3.10
importlib_metadata        8.6.1
ipykernel                 6.29.5
ipython                   8.34.0
ipywidgets                8.1.5
isoduration               20.11.0
jedi                      0.19.2
Jinja2                    3.1.6
joblib                    1.4.2
json5                     0.10.0
jsonpointer               3.0.0
jsonschema                4.23.0
jsonschema-specifications 2024.10.1
jupyter_client            8.6.3
jupyter_core              5.7.2
jupyter-events            0.12.0
jupyter-lsp               2.2.5
jupyter_server            2.15.0
jupyter_server_terminals  0.5.3
jupyterlab                4.3.6
jupyterlab_pygments       0.3.0
jupyterlab_server         2.27.3
jupyterlab_widgets        3.0.13
kiwisolver                1.4.8
Markdown                  3.7
markdown-it-py            3.0.0
MarkupSafe                3.0.2
matplotlib                3.10.1
matplotlib-inline         0.1.7
mdurl                     0.1.2
mistune                   3.1.3
mpmath                    1.3.0
multidict                 6.3.0
multiprocess              0.70.16
nbclient                  0.10.2
nbconvert                 7.16.6
nbformat                  5.10.4
nest-asyncio              1.6.0
networkx                  3.4.2
notebook_shim             0.2.4
numpy                     2.2.4
nvidia-cublas-cu12        12.4.5.8
nvidia-cuda-cupti-cu12    12.4.127
nvidia-cuda-nvrtc-cu12    12.4.127
nvidia-cuda-runtime-cu12  12.4.127
nvidia-cudnn-cu12         9.1.0.70
nvidia-cufft-cu12         11.2.1.3
nvidia-curand-cu12        10.3.5.147
nvidia-cusolver-cu12      11.6.1.9
nvidia-cusparse-cu12      12.3.1.170
nvidia-cusparselt-cu12    0.6.2
nvidia-nccl-cu12          2.21.5
nvidia-nvjitlink-cu12     12.4.127
nvidia-nvtx-cu12          12.4.127
overrides                 7.7.0
packaging                 24.2
pandas                    2.2.3
pandocfilters             1.5.1
parso                     0.8.4
peft                      0.15.2
pexpect                   4.9.0
pillow                    11.1.0
pip                       22.0.2
platformdirs              4.3.7
prometheus_client         0.21.1
prompt_toolkit            3.0.50
propcache                 0.3.1
protobuf                  3.20.3
psutil                    7.0.0
ptyprocess                0.7.0
pure_eval                 0.2.3
pyarrow                   19.0.1
pycparser                 2.22
Pygments                  2.19.1
pyparsing                 3.2.3
python-dateutil           2.9.0.post0
python-json-logger        3.3.0
pytz                      2025.2
PyYAML                    6.0.2
pyzmq                     26.3.0
referencing               0.36.2
regex                     2024.11.6
requests                  2.32.3
rfc3339-validator         0.1.4
rfc3986-validator         0.1.1
rich                      14.0.0
rpds-py                   0.24.0
safetensors               0.5.3
scikit-learn              1.6.1
scipy                     1.15.2
Send2Trash                1.8.3
sentencepiece             0.2.0
setuptools                59.6.0
shtab                     1.7.1
six                       1.17.0
sniffio                   1.3.1
soupsieve                 2.6
stack-data                0.6.3
sympy                     1.13.1
tensorboard               2.19.0
tensorboard-data-server   0.7.2
tensorboardX              2.6.2.2
terminado                 0.18.1
threadpoolctl             3.6.0
tinycss2                  1.4.0
tokenizers                0.21.1
tomli                     2.2.1
torch                     2.6.0
torchaudio                2.6.0
torchvision               0.21.0
tornado                   6.4.2
tqdm                      4.67.1
traitlets                 5.14.3
transformers              4.50.3
triton                    3.2.0
trl                       0.15.2
typeguard                 4.4.2
types-python-dateutil     2.9.0.20241206
typing_extensions         4.13.0
tyro                      0.9.18
tzdata                    2025.2
unsloth                   2025.3.19
unsloth_zoo               2025.3.17
uri-template              1.3.0
urllib3                   2.3.0
wcwidth                   0.2.13
webcolors                 24.11.1
webencodings              0.5.1
websocket-client          1.8.0
Werkzeug                  3.1.3
wheel                     0.45.1
widgetsnbextension        4.0.13
xformers                  0.0.29.post3
xxhash                    3.5.0
yarl                      1.18.3
zipp                      3.21.0

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

```
import torch
from peft import AutoPeftModelForSequenceClassification, PeftConfig, PeftModel
from transformers import AutoTokenizer, AutoModelForSequenceClassification

reward_lora_path = "/trains/models/train_20250415_142135"
label_names = ["score"]
id2label = {i: label for i, label in enumerate(label_names)}
label2id = {label: i for i, label in enumerate(label_names)}

peft_config = PeftConfig.from_pretrained(reward_lora_path)

peft_model = AutoPeftModelForSequenceClassification.from_pretrained(
    reward_lora_path, 
    config=peft_config,
    num_labels=len(label_names),
    id2label=id2label,
    label2id=label2id,
    torch_dtype=torch.bfloat16, 
    device_map="auto"
)

peft_model_2 = AutoPeftModelForSequenceClassification.from_pretrained(
    reward_lora_path, 
    config=peft_config,
    num_labels=len(label_names),
    id2label=id2label,
    label2id=label2id,
    torch_dtype=torch.bfloat16, 
    device_map="auto"
)

tokenizer = AutoTokenizer.from_pretrained(peft_config.base_model_name_or_path)
merged_model = peft_model_2.merge_and_unload(progressbar=True, safe_merge=True)

messages = [
    {"role": "user", "content": "what is 12*12?"},
    {"role": "assistant", "content": "144"},
]

text_encoded = tokenizer.apply_chat_template(messages, return_dict=True, return_tensors="pt", tokenize=True).to("cuda:0")
with torch.no_grad():
    score = peft_model(**text_encoded)
    score2 = merged_model(**text_encoded)
    print(f"peft model score:{score}")
    print(f"merged model score:{score2}")

```

得到结果如下:
```
peft model score:SequenceClassifierOutputWithPast(loss={'logits': tensor([[0.7422]], device='cuda:0', dtype=torch.bfloat16), 'past_key_values': <transformers.cache_utils.DynamicCache object at 0x7f37701ef9d0>}, logits=tensor([[0.7422]], device='cuda:0', dtype=torch.bfloat16), past_key_values=<transformers.cache_utils.DynamicCache object at 0x7f37701ef9d0>, hidden_states=None, attentions=None)
merged model score:SequenceClassifierOutputWithPast(loss={'logits': tensor([[0.4102]], device='cuda:0', dtype=torch.bfloat16), 'past_key_values': <transformers.cache_utils.DynamicCache object at 0x7f37701efe20>}, logits=tensor([[0.4102]], device='cuda:0', dtype=torch.bfloat16), past_key_values=<transformers.cache_utils.DynamicCache object at 0x7f37701efe20>, hidden_states=None, attentions=None)

```

Before merging, the score was 0.7442 (this meets the expectation). After merging, the score is: 0.4102 (does not meet the expectation).

### Expected behavior

It is expected that the results of model inference should be similar or the same before and after merging.

**Comment by weiminw** [comment]
补充, adapter_config.json:

```
{
  "alpha_pattern": {},
  "auto_mapping": null,
  "base_model_name_or_path": "/models/Qwen2.5-3B-Instruct",
  "bias": "none",
  "corda_config": null,
  "eva_config": null,
  "exclude_modules": null,
  "fan_in_fan_out": false,
  "inference_mode": true,
  "init_lora_weights": true,
  "layer_replication": null,
  "layers_pattern": null,
  "layers_to_transform": null,
  "loftq_config": {},
  "lora_alpha": 512,
  "lora_bias": false,
  "lora_dropout": 0.2,
  "megatron_config": null,
  "megatron_core": "megatron.core",
  "modules_to_save": [
    "score",
    "classifier",
    "score",
    "classifier",
    "score"
  ],
  "peft_type": "LORA",
  "r": 256,
  "rank_pattern": {},
  "revision": null,
  "target_modules": [
    "q_proj",
    "v_proj",
    "up_proj",
    "down_proj"
  ],
  "task_type": "SEQ_CLS",
  "trainable_token_indices": null,
  "use_dora": false,
  "use_rslora": false
}
```

**Comment by BenjaminBossan** [comment]
Could you please provide more information (preferably in English) about whether you used quantization on the base model (e.g. bitsandbytes 4bits)?

**Comment by weiminw** [comment]
> Could you please provide more information (preferably in English) about whether you used quantization on the base model (e.g. bitsandbytes 4bits)?

No, I don't use quantization model as base model, The base model is  Qwen2.5-3B-Instruct (bfloat16). the trained lora (trained by peft) saved as float32 default.

**Comment by BenjaminBossan** [comment]
Thanks for the info. I could reproduce a similar result to what you found. Here is a standalone reproducer:

```python
import os
import sys

import torch
from transformers import AutoTokenizer, AutoModelForSequenceClassification
from peft import AutoPeftModelForSequenceClassification, LoraConfig, PeftConfig, get_peft_model

torch.manual_seed(42)

reward_lora_path = "/tmp/peft/qwen2.5-3b-instruct"
model_id = "Qwen/Qwen2.5-3B-Instruct"
dtype = torch.bfloat16
#dtype = torch.float32

if not os.path.exists(reward_lora_path):
    model = AutoModelForSequenceClassification.from_pretrained(model_id, torch_dtype=dtype, num_labels=1)
    model = get_peft_model(model, peft_config=LoraConfig(task_type="SEQ_CLS", init_lora_weights=False))
    model.save_pretrained(reward_lora_path)
    print(f"Model saved to {reward_lora_path}")
    sys.exit(0)

label_names = ["score"]
id2label = {i: label for i, label in enumerate(label_names)}
label2id = {label: i for i, label in enumerate(label_names)}

peft_config = PeftConfig.from_pretrained(reward_lora_path)

peft_model = AutoPeftModelForSequenceClassification.from_pretrained(
    reward_lora_path,
    config=peft_config,
    num_labels=len(label_names),
    id2label=id2label,
    label2id=label2id,
    torch_dtype=dtype,
    device_map="auto",
)

peft_model_2 = AutoPeftModelForSequenceClassification.from_pretrained(
    reward_lora_path,
    config=peft_config,
    num_labels=len(label_names),
    id2label=id2label,
    label2id=label2id,
    torch_dtype=dtype,
    device_map="auto",
)

tokenizer = AutoTokenizer.from_pretrained(peft_config.base_model_name_or_path)
merged_model = peft_model_2.merge_and_unload(progressbar=True, safe_merge=True)

messages = [
    {"role": "user", "content": "what is 12*12?"},
    {"role": "assistant", "content": "144"},
]

text_encoded = tokenizer.apply_chat_template(messages, return_dict=True, return_tensors="pt", tokenize=True).to("cuda:0")
with torch.no_grad():
    score = peft_model(**text_encoded)
    score2 = merged_model(**text_encoded)
    print(f"peft model score:{score}")
    print(f"merged model score:{score2}")
```

(run it twice, first to create the PEFT checkpoint, then the 2nd time to load it and perform inference)

What I got is `2.3750` vs `2.4219`. The difference stems from the fact that the base model has half precision `torch.bfloat16` dtype, whereas the LoRA weights have full precision `torch.float32` dtype. When they are merged into the base model, they have to be downcast to `torch.bfloat16`, which introduces the error. In my snippet above, when I set `dtype = torch.float32`, I get the exact same results: `-3.2696` vs `-3.2696`.

If you cannot work with float32 and if you need the results to be closer after merging, there is one thing you could try: When you first create the PEFT model with `get_peft_model`, you can pass `autocast_adapter_dtype=False`. This should result in the PEFT weights being loaded in `torch.bfloat16` as well. Perhaps this will help reducing the error after merging, but it's not guaranteed.
